baseline/pytorch/tagger/model.py

```diff
@@ -136,83 +129,358 @@ def make_input(self, batch_dict, perm=False, numpy_to_tensor=False):
+    """Class defining a typical flow for taggers.  Most taggers should extend this class
+
+    This class provides the model base for tagging by providing specific hooks for each phase.  There are
+    3 basic operations identified in this class:
+
+    1. embed
+    2. encode (transduction)
+    3. decode
+
+    There is an `init_* method for each of this phases, allowing you to
+    define and return a custom layer.
+
+    The actual forward method is defined as a combination of these 3 steps, which includes a
+    projection from the encoder output to the number of labels.
+
+    Decoding in taggers refers to the process of selecting the best path through the labels and is typically
+    implemented either as a constrained greedy decoder or as a CRF layer
+    """
+    def __init__(self):
+        """Constructor"""
+        super().__init__()
+
+    def init_embed(self, embeddings: Dict[str, TensorDef], **kwargs) -> BaseLayer:
+        """This method creates the "embedding" layer of the inputs, with an optional reduction
+
+        :param embeddings: A dictionary of embeddings
+
+        :Keyword Arguments: See below
+        * *embeddings_reduction* (defaults to `concat`) An operator to perform on a stack of embeddings
+
+        :return: The output of the embedding stack followed by its reduction.  This will typically be an output
+          with an additional dimension which is the hidden representation of the input
+        """
+        return EmbeddingsStack(embeddings, self.pdrop, reduction=kwargs.get('embeddings_reduction', 'concat'))
+
+    def init_encode(self, input_dim, **kwargs) -> BaseLayer:
+        """Provide a layer object that represents the `encode` phase of the model
+        :param input_dim: The hidden input size
+        :param kwargs:
+        :return: The encoder
+        """
+
+    def init_decode(self, **kwargs) -> BaseLayer:
+        """Define a decoder from the inputs
+
+        :param kwargs: See below
+        :keyword Arguments:
+        * *crf* (``bool``) Should we create a CRF as the decoder
+        * *constraint_mask* (``tensor``) A constraint mask to apply to the transitions
+        * *reduction* (``str``) How to reduce the loss, defaults to `batch`
+        :return: A decoder layer
+        """
+        use_crf = bool(kwargs.get('crf', False))
+        constraint_mask = kwargs.get('constraint_mask')
+        if constraint_mask is not None:
+            constraint_mask = constraint_mask.unsqueeze(0)
+
+        if use_crf:
+            decoder = CRF(len(self.labels), constraint_mask=constraint_mask, batch_first=True)
+        else:
+            decoder = TaggerGreedyDecoder(
+                len(self.labels),
+                constraint_mask=constraint_mask,
+                batch_first=True,
+                reduction=kwargs.get('reduction', 'batch')
+            )
+        return decoder
+
+    def init_embed(self, embeddings: Dict[str, TensorDef], **kwargs) -> BaseLayer:
```

Comment:
You have these `init_(embed|encode|decode)` methods duplicated

Reply:
thanks